doc/tsung.1
.\" auto-generated by docbook2man-spec from docbook-utils package
.TH "TSUNG" "1" "January 2004" "" ""
.SH NAME
tsung \- A distributed multi-protocol load testing tool.
.SH SYNOPSIS
.sp
\fBtsung\fR [ \fB-f configuration file\fR ]  [ \fB-l log dir\fR ]  [ \fB-m filename\fR ]  [ \fB-r command\fR ]  [ \fB-v\fR ]  [ \fB-h\fR ]  [ \fBstart|stop|debug|status\fR ] 
.SH "DESCRIPTION"
.PP
\fBtsung\fR is a distributed load testing
tool. It is protocol-independent and can currently be used to
stress and benchmark HTTP, WebDAV, LDAP, PostgreSQL, MySQL and Jabber/XMPP servers.
.PP
It simulates user behaviour using an XML description file,
reports many measurements in real time (statistics can be
customized with transactions, and graphics generated using
gnuplot). 
.PP
For HTTP, it supports 1.0 and 1.1, has a proxy mode to
record sessions, supports GET and POST methods, Cookies, and Basic
WWW-authentication. It also has support for SSL. 
.PP
Several config examples can be found in
\fI/usr/share/doc/tsung/examples/\fR\&.
.TP
\fBstart\fR
start tsung load testing
.TP
\fBdebug\fR
start tsung with an interactive erlang shell
.TP
\fBstop\fR
stop tsung
.TP
\fBstatus\fR
print current status of a running instance of tsung (must be run on the controller host)
.SH "MANUAL"
.PP
A manual should be available at
\fI/usr/share/doc/tsung/user_manual.html\fR\&. It
is also available online at
.sp
.RS
.sp
.nf
http://tsung.erlang-projects.org/user_manual.html
.sp
.fi
.RE
.sp
.SH "OPTIONS"
.TP
\fB-f filename\fR
specifies the configuration file to use. The default
file name is \fI~/.tsung/tsung.xml\fR\&. Use - for standard input
.TP
\fB-l logdir\fR
Specifies the log directory to use. The default log dir name is \fI~/.tsung/log/YYYYMMDD-HHMM/\fR
.TP
\fB-m monitoring_file\fR
Specifies the monitoring log file name to use. The
default log file name
is \fItsung.log\fR\&. Use - for standard output
.TP
\fB-r command\fR
Specifies an alternative to ssh (rsh for ex.) for starting a slave node on a remote host
.TP
\fB-i id\fR
set controller id (default is empty). Needed to start
several controllers on the same host.
.TP
\fB-F\fR
Use long names for erlang nodes (FQDN)
.TP
\fB-m\fR
Enable erlang smp on client nodes
.TP
\fB-v\fR
Show version
.TP
\fB-h\fR
Show usage
.SH "BUGS"
.PP
Please reports bugs to the mailing list
<tsung-users@process-one.net>, see
.sp
.RS
.sp
.nf
https://lists.process-one.net/mailman/listinfo/tsung-users
.sp
.fi
.RE
.sp
for
archives.
.SH "SEE ALSO"
.PP
\fBerlang\fR(3)
.SH "AUTHORS"
.PP
\fBTsung\fR is written by Nicolas Niclausse 
<nicolas@niclux.org>\&. Contributors list
is available in
\fI/usr/share/doc/tsung/CONTRIBUTORS\fR


CHANGES
1.4.0 -> 1.4.1 Minor bugfixes (13 Sep 2011)
Bugfix:
   * [TSUN-188] - munin plugin is not working in 1.4.0
   * [TSUN-189] - the controller VM is not used in some case
   * [TSUN-190] - pgsql recorder can record a connect request in an already connected session
1.3.3 -> 1.4.0 Major enhancements and bugfixes (5 Sep 2011)
Bugfix:
   * [TSUN-129] - regexp (defined in match or dynvars) can fail when chunk encoding is used.
   * [TSUN-150] - Munin monitoring broken by cpu stats config request
   * [TSUN-163] - Tsung doesn't detect subdomains.
   * [TSUN-166] - snmp monitoring does not work with erlang R14A
   * [TSUN-171] - maxnumber set in a phase is not always enforced
   * [TSUN-172] - auth sasl can't authenticate against ejabberd
   * [TSUN-178] - some characters can make url and headers rewriting fail in the recorder
   * [TSUN-179] - tsung generated message stanzas are not XMPP compliant
   * [TSUN-180] - file server crash if a dynamic substitution use it
   * [TSUN-182] - When many clients are configured with few static users, none of them are launched.
   * [TSUN-183] - tsung can stop too soon in some cases
   * [TSUN-184] - 'random_number' with start and end actually returns a number from start+1 to end
   * [TSUN-187] - Client IP scan is very slow on Linux; also uses obsolete "ifconfig"
Improvements:
   * [TSUN-54] - tsung is very slow when a lot of dynamic variables are set
   * [TSUN-96] - generating more than 64k connections from a single machine
   * [TSUN-106] - Add content-encoding support for dynvar extraction
   * [TSUN-123] - add option to read usernames from an external file for jabber
   * [TSUN-125] - use the new re module everywhere instead of regexp/gregexp
   * [TSUN-152] - Add "state_rcv" record as parameter to "get_message" function.
   * [TSUN-153] - dynvar used in match may contain regexp special characters
   * [TSUN-185] - handle postgresql extended protocol
New Features:
   * [TSUN-162] - add foreach tag (loop when a dyn_variable is a list)
   * [TSUN-164] - add a switch to allow light queries/replies logging
   * [TSUN-165] - add a way to synchronize users for all plugins.
   * [TSUN-167] - add do=dump option to matching
   * [TSUN-168] - thinktimes value could be dynamically generated with dyn_variable
   * [TSUN-181] - add option to simulate slow connections
1.3.2 -> 1.3.3 Minor bugfixes (17 Aug 2010)
Bugfix:
    * [TSUN-154] - parent proxy doesn't work anymore in 1.3.x (tested with 1.3.2 and 1.3.0).
    * [TSUN-155] - url substitution is broken in some cases
    * [TSUN-156] - Tsung not using sessions with low probabilities
    * [TSUN-157] - ssl doesn't work with erlang R14A
    * [TSUN-158] - failure when a proxy is used and an URL substitution is set
    * [TSUN-159] - HTTP cookies support is broken when a proxy is used
    * [TSUN-160] - tsung can sometimes hang at the beginning using distributed setup
    * [TSUN-161] - if statement not allowed in a transaction
1.3.1 -> 1.3.2 Major bugfixes and enhancements (14 Jun 2010)
Bugfix:
    * [TSUN-128] - Apostrophes cause string to convert to deep list in setdynvars with Erlang function.
    * [TSUN-130] - static users starting time is wrong
    * [TSUN-131] - tsung can stop too early when static users are used
    * [TSUN-132] - http cookies: accept domains equals to hostname with a leading "."
    * [TSUN-133] - proxy-recorder with SSL fails on large client packets (multiple TCP packets)
    * [TSUN-138] - when an error occured( for ex a timeout during a request) and a client exits, started transactions are not updated
    * [TSUN-140] - tsung does not honor the Proxy-Connection: keep-Alive or Connection: keep-Alive header if the proxy is HTTP/1.0
    * [TSUN-142] - http recorder can fail with https rewriting and chunked encoding
    * [TSUN-147] - UDP & bidi does not seem to work
    * [TSUN-148] - dynvar not found when used in match
    * [TSUN-149] - tsung doesn't work with Amazon Elastic load balancing
    * [TSUN-151] - tsung_stats.pl produces invalid *.gplot files
Improvements:
    * [TSUN-82]  - XMPP vhost support
    * [TSUN-127] - add ability tu use floats for thinktimes
    * [TSUN-139] - option to set random seed for launchers.
    * [TSUN-141] - add dynamic variable support for postgres
    * [TSUN-146] - New tsplot yfactor configuration support breaks most common configurations
New Features:
    * [TSUN-135] - add support for SASL ANONYMOUS and PLAIN authentication for XMPP
    * [TSUN-136] - add new plugin distributed testing of filesystem (nfs for ex), using a generic mode for executing erlang functions on clients nodes
    * [TSUN-137] - add a way to mix requests types inside a single session
    * [TSUN-143] - global time limit for the load test
    * [TSUN-145] - tsung should support json parsing for dynamic variable
1.3.0 -> 1.3.1 Major bugfixes and enhancements (9 Sep 2009)
Bugfix:
    * [TSUN-92] - the computation of the minimum for sample_counter is wrong
    * [TSUN-93] - maxnumber not respected if several clients are used
    * [TSUN-102] - dyn_variable configuration fails if variable name is not a valid erlang atom
    * [TSUN-103] - Network error handling in munin plugin
    * [TSUN-104] - tsung-plotter can't handle the os_mon statistics
    * [TSUN-108] - Cannot handle complicated dyn_var name
    * [TSUN-109] - Tsung status displays always phase one even if you have more than one phase
    * [TSUN-110] - Cookie header not present if the URL is dynamically generated by a previous redirection (302)
    * [TSUN-117] - Bug in HTTP: empty header can be generated in some case
    * [TSUN-118] - HTTPS proxy recorder: ts_utils:to_https incorrectly handles Content-Length for POST requests
    * [TSUN-119] - tsung can crash when reading empty values from a csv file
    * [TSUN-122] - same http cookie key with different domains don't work
Improvements:
    * [TSUN-47] - ts_mon can be a bottleneck during very high load testing

src/tsung-plotter/tsung/tsung.py
#! /usr/bin/python
# -*- coding: utf-8 -*-

#  Copyright: 2006 by Dalibo <dalibo.com>
#  Copyright: 2007 Dimitri Fontaine <dim@tapoueh.org>
#  Created: 2006 by Dimitri Fontaine <dim@tapoueh.org>
#
#  Modified: 2008 by Nicolas Niclausse
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
#
#  In addition, as a special exception, you have the permission to
#  link the code of this program with any library released under
#  the EPL license and distribute linked combinations including
#  the two.

"""
Python classes for tsung log data usage
 data: generic tsung data
  sample
  counter
  gauge

 log: tsung log file parser, using sample, gauge and counter classes

This package has its own configuration file where to associate tsung
stats names with tsung stat type.
"""

from ConfigParser import ConfigParser

# data are produced every 10 seconds, by default
# we read real used interval in log file (deduced from timestamps)
INTERVAL  = 10
SEPARATOR = "# stats: dump at "
PREFIX    = "stats: "

class data:
    """ Tsung logged data, to be specialized into sample and counter """
    def __init__(self):
        pass

    def get(self, name):
        """ get named statistic """
        if name not in self.__dict__.keys():
            return None

        # don't return any value when no measure have been taken
        # durung interval
        if self.count == 0:
            return None

        if name == 'count' :
            if self.interval > 0:
                return self.count / self.interval
            else:
                return 0

        return self.__dict__[name]

class sample(data):
    """ tsung sample stats """

    def __init__(self, interval = INTERVAL, values = [0, 0, 0, 0, 0]):
        """ init log data values """
        self.interval = interval

        self.count  = float(values[0])
        self.mean   = float(values[1])
        self.stdvar = float(values[2])
        self.min    = float(values[3])
        self.max    = float(values[4])
        self.gmean  = float(values[4])

    def __repr__(self):
        """ human readable output """
        return "sample: %s %s %s %s %s" % (self.value, self.mean,
                                           self.stdvar, self.min, self.max, self.gmean)

class counter(data):
    """ tsung counter stats """

    def __init__(self, interval = INTERVAL, values = [0, 0]):
        """ init log data values """
        self.interval = 1

        self.count      = float(values[0])
        self.totalcount = float(values[1])

    def __repr__(self):
        """ human readable output """
        return "counter: %s %s" % (self.value, self.totalcount)

class gauge(data):
    """ tsung gauge stats """

    def __init__(self, interval = INTERVAL, values = [0, 0]):
        """ init log data values """
        self.interval = 1

        self.count      = float(values[0])
        self.max        = float(values[1])

    def __repr__(self):
        """ human readable output """
        return "gauge: %s %s" % (self.count, self.max)

class TsungLog:
    """ Tsung logged data parser and representation

    data is a {timestamp, data} hash, with data either sample or
    counter instance.
    """

    def __init__(self, config_filename, filename):
        """ constructor """
        self.conffile = config_filename
        self.filename = filename
        self.types    = {}
        self.data     = {}
        self.unknown  = []

        self.configure()
        self.parse()

    def configure(self):
        """ read configuration file to associates stat types to stat names """

        config = ConfigParser()
        config.read(self.conffile)

        for s in config.sections():
            for (name, type) in config.items(s):
                if type == 'sample':
                    self.types[name] = sample()
                if type == 'counter':
                    self.types[name] = counter()
                if type == 'gauge':
                    self.types[name] = gauge()

    def parse(self):
        """ read data from self.filename """
        first_ts   = 0
        current_ts = 0
        record_ts  = 0
        import re

        for line in file(self.filename):
            # chomp \n
            line = line[:-1]

            if line.find(SEPARATOR) == 0:
                measure_ts = int(line[len(SEPARATOR):].strip())

                # interval in seconds between two records. As tsung
                # generates a reference line, sensible values won't
                # have to divide by 0
                interval = measure_ts - current_ts

                # now we change current_ts
                current_ts = measure_ts

                if first_ts == 0:
                    first_ts  = current_ts

                record_ts = current_ts - first_ts

            elif line.find("stats: ") == 0:
                array  = [v.strip() for v in line.split(" ")]
                name   = array[1]
                values = array[2:]

                if self.types.has_key(name):
                    data = self.types[name].__class__(interval, values)

                    if not self.data.has_key(record_ts):
                        self.data[record_ts] = {}

                    self.data[record_ts][name] = data

                else:
                    is_re = False
                    x = re.compile("^[\w\d]+$")
                    for k in self.types.keys():
                        if not re.match(x, k):
                            y = re.compile(k)
                            if re.match(y, name):
                                data = self.types[k].__class__(interval, values)
                                if not self.data.has_key(record_ts):
                                    self.data[record_ts] = {}
                                self.data[record_ts][name] = data
                                is_re = True
                                break

                    if name not in self.unknown and not is_re:
                        print 'WARNING: tsung %s data is not configured' % name
                        self.unknown.append(name)

    def stat(self, name, stat):
        """ returns a {timestamp: date_type} dict for given named statistic """
        ret = {}
        for ts, stats in self.data.items():
            if stats.has_key(name) and stats[name].get(stat) is not None:
                ret[ts] = stats[name].get(stat)

        return ret


if __name__ == '__main__':
    # some unit testing
    import sys

    config   = "stats.conf"
    logfile  = sys.argv[1]

    tsunglog = TsungLog(config, logfile)

    from pprint import pprint
    pprint(tsunglog.stat('request', 'mean'))
    pprint(tsunglog.stat('users', 'max'))
    pprint(tsunglog.stat('finish_users_count', 'count'))
    pprint(tsunglog.stat('200', 'count'))

